phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -614,38 +606,132 @@ private boolean isDeleteFamilyVersion(Mutation mutation) {
         return getMutationsWithSameTS(put, del);
     }
 
+    private void repairActualMutationList(List<Mutation> actualMutationList, List<Mutation> expectedMutationList)
```

Comment:
I am just thinking aloud here. Read repair has similar logic to generate the expected index mutation, but only for specific timestamp. Is it possible to leverage the code there or unify them?
One side question. From the code of repairIndexRows, it looks like opening a lot of scan. If there are multiple concurrent read repair happen (may be caused by some bug int the future or network issue), will it overload the hbase region server and cause the cluster not be able to recover from such network hitcup?

Reply:
This simulates the read repair given that the list of revisions for the actual and the expected (repaired form) mutations. It is much simpler than the read repair process in GlobalIndexChecker which works on one index row at a time and leverages the HBase scans on the actual table, which is not the case here. Also, GlobalIndexChecker also purges the invalid rows from HBase in addition rebuilding unverified rows. So, they are very different in nature and not sure how they can be unified.

Reply:
thanks